File: TP02Q10/jogador.cpp
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

#define NUMENTRADA 1000
#define TAMLINHA 1000
int comp;

typedef struct Jogador{
    int id, altura, peso, anoNascimento;
    char nome[100], universidade[100], cidadeNascimento[100], estadoNascimento[100];
}Jogador;

bool isFim(char s[]){
    return (strlen(s) >= 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}

void ler(Jogador *j, char player[]){
    j->id = atoi(strsep(&player,","));

    strcpy(j->nome, strsep(&player , ","));    

    j->altura = atoi(strsep(&player,","));    

    j->peso = atoi(strsep(&player,","));    

    strcpy(j->universidade, strsep(&player , ","));
    if(strcmp(j->universidade, "")==0)
        strcpy(j->universidade, "nao informado");

    j->anoNascimento = atoi(strsep(&player,",")); 
    
    strcpy(j->cidadeNascimento, strsep(&player , ","));
    if(strcmp(j->cidadeNascimento, "")==0)
        strcpy(j->cidadeNascimento, "nao informado");
    
    for(int i=0; i<strlen(player); i++){
        if(player[i]=='\n'){
           player[i]='\0';
        }   
    }
    strcpy(j->estadoNascimento, player);
    if(strcmp(j->estadoNascimento, "")==0)
        strcpy(j->estadoNascimento, "nao informado");
    
}

void imprimir(Jogador *i){
    printf("[%i ## %s ## %i ## %i ## %i ## %s ## %s ## %s]\n",i->id ,i->nome , i->altura, i->peso, i->anoNascimento, i->universidade, i->cidadeNascimento, i->estadoNascimento);
}

void log(long cronometro){
    FILE *arq;
    arq = fopen("677213_quicksort.txt","w");
    fprintf(arq,"677213\t%ld\t%d",cronometro,comp);
    fclose(arq);
}

void swap(int a, int b , Jogador array[]){
    Jogador aux = array[a];
    array[a] = array[b];
    array[b] = aux;
}

 void quicksort(int esq, int dir , Jogador array[]){
    int i = esq, j = dir;
    Jogador pivo = array[(dir+esq)/2];
    while (i <= j) {
        while (strcmp(array[i].estadoNascimento,pivo.estadoNascimento) < 0){ i++; comp++;}
        while (strcmp(array[j].estadoNascimento,pivo.estadoNascimento) > 0){ j--; comp++;}
        if (i <= j) {
            swap(i, j , array);
            i++;
            j--;
        }
    }
    if (esq < j)  
        quicksort(esq, j, array);
    if (i < dir)  
        quicksort(i, dir, array);
}

int main(){
    char entrada[NUMENTRADA][TAMLINHA];
    int numEntrada = 0;
    comp = 0;
    
    do{
        fgets(entrada[numEntrada],TAMLINHA, stdin);
    }while(isFim(entrada[numEntrada++]) == false);
    numEntrada--;
    
    
    char players[3923][200];
    int numLinha = 0;
    
    FILE * fPointer;
        fPointer = fopen("/tmp/players.csv", "r");   

        do{
            fgets(players[numLinha],200,fPointer);  
            numLinha++; 
        }while(numLinha<3923);
                

        fclose(fPointer);

        Jogador jogador[numEntrada];
        int pos[numEntrada];

        for(int i= 0; i < numEntrada; i++){
            ler(&jogador[i] , players[atoi(entrada[i])+1]);
        }

        long tempo = clock();
        quicksort(0, (numEntrada-1), jogador);
        tempo = clock() - tempo;
        log(tempo);
            for (int i = (numEntrada - 1); i > 0; i--) {
            for (int j = 0; j < i; j++) {
                if ((strcmp(jogador[j].nome,jogador[j + 1].nome) > 0) && (strcmp(jogador[j].estadoNascimento,jogador[j + 1].estadoNascimento) == 0)) {
                    Jogador a = jogador[j];
                    jogador[j]=jogador[j+1];
                    jogador[j+1]=a;
                }
            }
        }

        for(int i= 0; i < numEntrada; i++){
            imprimir(&jogador[i]);
        }
    return 0;
}

File: TP02Q13/Jogador.java
class Jogador{
    private int id, altura, peso, anoNascimento;
    private String nome, universidade, cidadeNascimento, estadoNascimento;
    private static Jogador[] array;
    
    public static boolean isFim(String s){
        return(s.length()>= 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
    }

    public Jogador(){
        this.id = 0;
        this.altura = 0;
        this.peso = 0;
        this.anoNascimento = 0;
        this.nome = "";
        this.universidade = "";
        this.cidadeNascimento = "";
        this.estadoNascimento = "" ;   
    } 
    
    public Jogador(int id, int altura, int peso, int anoNascimento, String nome, String universidade, String cidadeNascimento, String estadoNascimento){
        this.id = id;
        this.altura = altura;
        this.peso = peso;
        this.anoNascimento = anoNascimento;
        this.nome = nome;
        this.universidade = universidade;
        this.cidadeNascimento = cidadeNascimento;
        this.estadoNascimento = estadoNascimento;  
    }

    public void setId(int id){
        this.id = id;
    }

    public int getId(){
        return this.id;
    }

    public void setAltura(int altura){
        this.altura = altura;
    }

    public int getAltura(){
        return this.altura;
    }

    public void setPeso(int peso){
        this.peso = peso;
    }

    public int getPeso(){
        return this.peso;
    }

    public void setAnoNascimento(int anoNascimento){
        this.anoNascimento = anoNascimento;
    }

    public int getAnoNascimento(){
        return this.anoNascimento;
    }

    public void setNome(String nome){
        this.nome = nome;
    }

    public String getNome(){
        return this.nome;
    }

    public void setUniversidade(String universidade){
        this.universidade = universidade;
    }

    public String getUniversidade(){
        return this.universidade;
    }

    public void setCidadeNascimento(String cidadeNascimento){
        this.cidadeNascimento = cidadeNascimento;
    }

    public String getCidadeNascimento(){
        return this.cidadeNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento){
        this.estadoNascimento = estadoNascimento;
    }

    public String getEstadoNascimento(){
        return this.estadoNascimento;
    }

    public Jogador clone(){
        return new Jogador(this.id, this.altura, this.peso, this.anoNascimento, this.nome, this.universidade, this.cidadeNascimento, this.estadoNascimento);
    }
    
    public void ler(String s){
        String[] array = s.split(",");   

        this.id = Integer.parseInt(array[0]);
        this.nome = array[1];
        this.altura = Integer.parseInt(array[2]);
        this.peso = Integer.parseInt(array[3]);
            if(array[4].length() == 0)
                this.universidade = "nao informado";
            else
                this.universidade = array[4];
        this.anoNascimento = Integer.parseInt(array[5]);
            if(array.length == 8){
                this.cidadeNascimento = array[6];
                this.estadoNascimento = array[7];
            }
            else{
                this.cidadeNascimento = "nao informado";
                this.estadoNascimento = "nao informado";
            }  
    }

    public void imprimir(){
        System.out.println("["+getId()+" ## "+getNome()+" ## "+getAltura()+" ## "+getPeso()+" ## "+getAnoNascimento()+" ## "+getUniversidade()+" ## "+getCidadeNascimento()+" ## "+getEstadoNascimento()+"]");
    }

    public static void log(int comp , long tempo){
        Arq.openWrite("677213_countingsort.txt");

        Arq.print("677213"+'\t'+tempo+'\t'+comp);

        Arq.close();
    }
 
    public static void intercalar(int esq, int meio, int dir){
        int i, j, k;
        
        //Definir tamanho dos dois subarrays
        int n1 = (meio +1)- esq;
        int n2 = dir - meio;
        
        Jogador[] a1 = new Jogador[n1 + 1];
        Jogador[] a2 = new Jogador[n2 + 1];

        //Inicializar primeiro subarray
        for(i = 0; i < n1; i++){
            a1[i] = array[esq+i];
        }
 
       //Inicializar segundo subarray
       for(j = 0; j < n2; j++){
          a2[j] = array[(meio+1)+j];
       }
 
       //Sentinela no final dos dois arrays
       Jogador sentinela = new Jogador();
       sentinela.setUniversidade("zzzzzzzzzzzzzzzzzzzzzzZzZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ");
       a1[i] = a2[j] = sentinela;
 
       //Intercalacao propriamente dita
       for(i = 0 , j = 0, k = esq; k <= dir; k++){
          array[k] = ((a1[i].getUniversidade().compareTo(a2[j].getUniversidade()) < 0) ? a1[i++] : a2[j++]);
       }
    }
    
        public static void mergesort(int esq , int dir) {
        if (esq < dir){
           int meio = (esq + dir) / 2;
           mergesort(esq, meio);
           mergesort(meio + 1, dir);
           intercalar(esq, meio, dir);
        }
     }

public static void main(String[] args){
    String[] entrada = new String[1000];
    String[] nomes = new String[1000];
    int numEntrada = 0;
    int numNomes = 0;
    

    do{
        entrada[numEntrada] = MyIO.readLine();
    }while(isFim(entrada[numEntrada++]) == false);
    numEntrada--;

    Jogador[] jogador = new Jogador[numEntrada];
    array = new Jogador[numEntrada];
    String[] players = new String [5000];

    Arq.openRead("/tmp/players.csv");// /tmp/

    for (int i = 0 ; i<3923; i++){
        players[i]= Arq.readLine();
    } 
    
    Arq.close();

    int x = 0;
    for(int i = 0 ;  i< numEntrada ; i++){
        x = Integer.parseInt(entrada[i]);
        jogador[i] = new Jogador();
        jogador[i].ler(players[x+1]);
        //jogador[i].imprimir();
    }
    array = jogador;
    
    mergesort(0, jogador.length - 1);

    jogador = array;

        for (int i = (numEntrada - 1); i > 0; i--) {
            for (int j = 0; j < i; j++) {
                if (jogador[j].getNome().compareTo(jogador[j + 1].getNome()) > 0 && jogador[j].getUniversidade().compareTo(jogador[j + 1].getUniversidade()) == 0) {
                    Jogador aux = jogador[j];
                    jogador[j]=jogador[j+1];
                    jogador[j+1]=aux;
                }
            }
        }

        // Impressao
        for(int i = 0 ;  i< numEntrada ; i++){
           jogador[i].imprimir();
        }
     

    } 
}